### ISSUE TITLE
[java] UnnecessaryCast: False positives related to generics

### ISSUE THREAD

**Issue by UncleOwen** [pmd/pmd#6943]
**Affects PMD Version:** 7.27.0-SNAPSHOT

**Rule:** [UnnecessaryCast](https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#unnecessarycast)

**Description:**

While trying to add UnnecessaryCast to the dogfood ruleset, I encountered the following FPs:

**Code Sample demonstrating the issue:**

```java
String valueString = ((PropertyDescriptor) propertyDescriptor).serializer().toString(value);
```
https://github.com/pmd/pmd/blob/main/pmd-core/src/main/java/net/sourceforge/pmd/properties/AbstractPropertySource.java#L172

```java
ShadowChainBuilder<S, ?>.ResolverBuilder builder = (ShadowChainBuilder<S, ?>.ResolverBuilder) classes.new ResolverBuilder();
```
https://github.com/pmd/pmd/blob/main/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/table/internal/JavaResolvers.java#L525

**Expected outcome:**

In each case, PMD reports a violation, but that's wrong. That's a false positive. Removing the cast results in a compile error.

**Running PMD through:** *Maven*
